File: Python/Problem4.py
from queue import Queue
from typing import List, Tuple
import copy

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        n,m = len(grid), len(grid[0])
        if n==0:
            return 0
        q = Queue()
        vis = [[0] * m for _ in range(n)]
        orangesToRot = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 2 and not vis[i][j]:
                    q.put(((i,j),0))
                    vis[i][j] = 2
                if grid[i][j] == 1:
                    orangesToRot += 1

        if orangesToRot == 0:
            return 0

        delrow = [-1, 0, 1, 0]
        delcol = [0, -1, 0, 1]
        rottedOranges = 0
        time = 0
        while not q.empty():
            (row, col), t = q.get()
            time = max(time, t)

            for i in range(4):
                newr = delrow[i] + row
                newc = delcol[i] + col
                if 0<=newr<n and 0<=newc<m and grid[newr][newc]==1 and not vis[newr][newc]:
                    vis[newr][newc] = 2
                    q.put(((newr,newc), t+1))
                    rottedOranges += 1


        if rottedOranges != orangesToRot:
            return -1

        return time

    def newGrid(self, grid: List[List[int]]) -> List[List[int]]:
        n,m = len(grid), len(grid[0])
        if n==0:
            return [[]]
        q = Queue()
        # vis = [grid[:] for row in grid] # Create a deep copy of the input grid
        vis = copy.deepcopy(grid)
        # print("vis before:", vis)
        orangesToRot = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 2:
                    q.put(((i,j),0))
                if grid[i][j] == 1:
                    orangesToRot += 1

        if orangesToRot == 0:
            return vis

        delrow = [-1, 0, 1, 0]
        delcol = [0, -1, 0, 1]
        rottedOranges = 0
        time = 0
        while not q.empty():
            # print("Queue contents:", q.queue)
            (row, col), t = q.get()
            time = max(time, t)
            # print("row:", row, "col:", col, "t:", t)

            for i in range(4):
                newr = delrow[i] + row
                newc = delcol[i] + col
                # print("newr:", newr, "newc:", newc)
                if 0<=newr<n and 0<=newc<m and grid[newr][newc]==1 and vis[newr][newc]==1:
                    vis[newr][newc] = 2
                    q.put(((newr,newc), t+1))
                    rottedOranges += 1


        if rottedOranges != orangesToRot:
            return [[]]

        return vis


if __name__ == "__main__":
    grid = [[0, 1, 2], [0, 1, 2], [2, 1, 1]]

    print("original grid: ")
    for row in grid:
        print(" ".join(map(str, row)))

    print("---------------------")
    obj = Solution()
    ans = obj.orangesRotting(grid)
    newGrid = obj.newGrid(grid);

    print("new grid: ")
    for row in newGrid:
        print(" ".join(map(str, row)))

    print("time to rot all oranges: ", ans)

File: JAVA/Problem4.java
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;

class Solution{
  public int orangesRotting(int[][] grid){
    if(grid == null || grid.length == 0) return 0;
    int n = grid.length;
    int m = grid[0].length;
    Queue<int[]> q = new LinkedList<>();
    int orangesToRot = 0;
    for(int i=0; i<n; i++){
      for(int j=0; j<m; j++){
        if(grid[i][j]==2){
          q.offer(new int[]{i,j});
        }
        if(grid[i][j]!= 0){
          orangesToRot++;
        }
      }
    }
    if (orangesToRot == 0) {
        // No fresh oranges to rot, return 0
        return 0;
    }
    int delrow[] = {-1, 0, 1, 0};
    int delcol[] = {0, -1, 0, 1};
    int orangesRotted = 0;
    int time = 0;
    while(!q.isEmpty()){
      int qsize = q.size();
      orangesRotted += qsize;

      for(int t=0; t<qsize; t++){
        int[] arr = q.poll();
        for(int i=0; i<4; i++){
          int newr = delrow[i] + arr[0];
          int newc = delcol[i] + arr[1];
          if(newr<0 || newc<0 || newr>=n || newc>=m || grid[newr][newc]==0 || grid[newr][newc]==2)
            continue;

          grid[newr][newc] = 2;
          q.offer(new int[]{newr, newc});

        }
      }
      if(q.size() != 0){
        time++;
      }
    }
    return orangesToRot == orangesRotted ? time : -1;
  }

  public static void main(String[] args){

    int[][] grid = {{0,1,2},{0,1,2},{2,1,1}};
    System.out.println("original grid: ");
    for(int i=0; i<grid.length; i++){
      for(int j=0; j<grid[i].length; j++){
        System.out.print(grid[i][j] + " ");
      }
      System.out.println();
    }

    System.out.println("-------------------------------");

    Solution obj = new Solution();
    int ans = obj.orangesRotting(grid);

    System.out.println("new grid: ");
    for(int i=0; i<grid.length; i++){
      for(int j=0; j<grid[i].length; j++){
        System.out.print(grid[i][j] + " ");
      }
      System.out.println();
    }
    System.out.println("Time to take to rot all oranges: " + ans);
  }
}
